### ISSUE TITLE
Testing platform-specific code (-p on GCC in particular)

### ISSUE THREAD

**Issue by arstrube** [cpputest/cpputest#460]
I have felt uneasy before about platform-specific code never receiving any direct testing.

I am not sure that #ifdef'ing these tests into the main test executables is the way to go, either.

I am about to add some more code (that needs tests) to the platform-specific run-in-separate process function. I propose to look at the testing issue in a wider context and to come up with a concept how platform-specific code can be tested.

Naturally, we will focus on what can be tested with GCC and Travis-CI. It might make sense to create a separate test executable (so as to not unnecessarily bloat the Travis build). It might also make sense to put the tests into their respective platform-specific locations.

Contributors for other platforms could contribute tests to be run on those platforms only.

What are your thoughts?

**Comment by arstrube** [comment]
Any ideas yet?

**Comment by basvodde** [comment]
Nope :) It is hard as it depends on the platform :)

**Comment by Andne** [comment]
Aren't the platform-specific functions supposed to provide a common interface for the rest of the system?  That implies to me that there should be a common behavior that can the tests can check against regardless of the platform.

**Comment by arstrube** [comment]
@Andne Common interface, yes. Common behavior, only if we manage to provide implementations for at least the major platforms (Unix/MacOs, Cygwin, MinGW, Visual Studio). On some platforms, there is an empty implementation (or one that just prints a message).

So, if I use the common interface for testing, we will create some more conditional defines and IGNORE_TEST.

If, however, the test was part of the platform specific stuff, I could test at that level and not worry that e.g. MinGW currently doesn't implement that behavior.

**Comment by Andne** [comment]
I think that the platform-specific code could probably be split into two parts - core required functions, that must have an expected behavior, and secondary functions that assist functionality that isn't required and may not work on all platforms.

The first group would include things like setjmp, longjmp, printf, etc... and hopefully any tests would have the same behavior across all platforms.

The second list would include things like the separate process logic and wouldn't be possible on all platforms (especially when embedded gets mixed in) and so would require different tests at times.

**Comment by arstrube** [comment]
Yes. I guess I am talking about the second list here. Any ideas would be welcome :)

**Comment by arstrube** [comment]
I looked at the tests for Mutex. Unfortunately, they are not helpful here because I need to test the platformspecific function itself, no stubbing it is not really helpful.

The tests should verify that
- recursion does not occur (the bug I already fixed
- various messages are what they are supposed to be (the code I am yet to submit)

Questions: 
1. Where _exactly_ do these tests go?
2. Do we work with conditional compilation?
3. If so, is `__MINGW__` still the right criterion? After all, it would appear that MinGW _can_ have PTHREAD support, so maybe the real criterion is whether PTHEAD support is available?

When we do have PThread support, nothing needs to be `#ifdef`ed away for MinGW/GCC.

It would be really helpful to get some concrete feedback here.

**Comment by Andne** [comment]
I played around with different builds of MinGW and haven't found a way to really detect which build someone has.  I think what I have found though is that it's technically not a matter of which toolchain is being used, it matters if the pthread library exists or not.

On Unix platforms, it seems to always exist, but there's a -pthread option that does something, though it seems to mostly be for the linker.

With the MinGW-w64 builds for windows, both posix and win32 thread builds have libpthread in them, so I think either one can use pthreads, the real difference is what the toolchain itself uses.  I saw a couple posts online that seemed to imply that this needs to agree with the library to avoid issues during exception handling, but other posts that disagreed with this.

My interpretation of all of this:
1. All you need for pthreads is pthread.h and the library
2. Mingw-builds of mingw-w64 (the versions that I tend to use) include pthread.h and the library
3. So can we just assume that MinGW has pthreads and be done with it?  Maybe not due to older/other builds of MinGW likely not including libwinpthreads.  Otherwise, the only detection I can find is to see if pthread.h exists or not - maybe ok for autotools, not so sure about cmake.
4. Wouldn't it be nice if there was a 'standard' implementation of MinGW?

**Comment by basvodde** [comment]
I'm not quite sure where these test would fit. I'd suggest to just start with conditional complication and when it becomes ugly, we'll fix it in a different way.

Related to pthreads, this shouldn't be a platform #define but it should be set in the configure script, like much other things. That isn't used correctly now (same for e.g. 64 bits)

**Comment by arstrube** [comment]
Should we fix the configure script first, then? The thing is, right now, I am not quite sure what to base conditional compilation on (it would have to be `!defined(__MINGW32__) ` or some such thing right now).

**Comment by arstrube** [comment]
I assume the configure script would be able to figure out the presence of pthread, as well as the target architecture (e.g. ILP64, IL32P64 and so on) ?

**Comment by basvodde** [comment]
Yes, but the idea behind configure and automake it so test for capability and not test for a particular architecture

**Comment by arstrube** [comment]
Hmm. So pthread support would be a capability.

As for longs... as far as I know, configure tests for the availability of types such as `uint64_t`. This could be tested for. Generally, CppUTest seems to avoid <cstdint> though. Generally though, 'modern' platform independent code is written using standard int (or some equivalent), leaving the definition of these types to the compiler. E.g. a Windows compiler would typedef `uint64_t` as `unsigned long long`.

**Comment by arstrube** [comment]
I believe `uint64_t` would work on all architectures, even most 32 bit architectures. If the "feature" uint64_t` were not available, (which would never happen on any kind of 64 bit system), then 64 bit pointer couldn't be cast to `uint64_t`, but neither would that be necessary.

**Comment by Andne** [comment]
I think you're right, using the stdint types should remove one or two of the platform-specific checks that are being performed today.  I suspect that part of the reason those typedefs aren't used is because of the desire to support environments that don't have the C standard library (only one I know is Linux kernel development), but I would think that those should still have something, so could we just let the typedefs be in a platform-specific header for that set of development instead?

**Comment by basvodde** [comment]
Yes, the platform dependent platform independent typedefs are still avoided. So far, we'd like to keep that ;)

**Comment by Andne** [comment]
I think that not using typedefs for types is part of what's causing the need to have some tests enable/disable based on the environment, such as what was discussed in #472.  I understand if you want to avoid using the stdint typedefs in order to avoid usage of libc headers in the case of environments that have them.  In that case, should there be a set of typedefs specific to CppUTest that can be mapped to the appropriate types instead then?

**Comment by basvodde** [comment]
So far, I've always wanted to avoid it... having nothing but negative experiences for lots of typedef-codebases. What is your concrete proposal?

**Comment by arstrube** [comment]
I do understand the idea behind stdint typdefs kind of takes the opposite approach of how C was originally designed to be platform-independent. I believe this came about by the need to know e.g. the size of the underlying types. In the embedded world, it is considered unsafe to work with types whose size is, potentially, changing.

Interestingly though, all embedded projects I have been part of have tried to avoid stint.h as well, creating their own type like sint16 etc. instead. This usually works fine until they need to include code from some other supplier that uses their own brand of typedefs, which are different or - worse even - the same.

As CppUTest is following the philosophy of 'original' C, I don't think turning around 180 degrees and using specific width ints everywhere (with or without stdint.h) would be a good idea.

How about just using it for the one type of int that is giving us trouble, 64bit ints? A conditional typedef (based on the configure process), like u64 could be really helpful here, in my opinion.

On the other hand, I don't quite understand why we don't simply use `long long` on Windows systems. This would be more in line with the current philosophy behind CppUTest and limited to the Windows platform. It wouldn't get rid of conditional compilation, but at least it would provide valid output for 64 bit pointers on Windows.

Whereas, going the other route, and always casting pointers to 64bit (u64), while it might work, would produce strange output on 32 bit systems, like `0x00000000045243FA`

**Comment by arstrube** [comment]
Just noticed how totally off-topic I went with this discussion :-(.

Actually, I have some code ready to submit for running in a separate process, with improved failure messages. What I am looking for is some guidance with how to add tests for it (I ran my own in a separate project). The thing I'd need to do for testing is roughly:

- Run tests in a separate process (not by running a separate executable with -p). 
- Make the test cases crash for a variety of reasons
- Intercept the results / messages and check them against expected results / messages

I have seen some similar test cases (that test certain output strings) but so far, non that I could directly model.

My own tests look like this, but they require manual checking at this point. So they are not really good unit tests.

```c++
TEST_GROUP(Separate_Process) {};

TEST(Separate_Process, thisTestIsStillRunAndSucceeds) {
    LONGS_EQUAL(1, 1);
}

TEST(Separate_Process, thisTestIsStillRunAndFails) {
    LONGS_EQUAL(3, 1);
}

TEST(Separate_Process, thisTestIsStillRunAndCrashesWithDivisionByZero) {
	int bar = 0;
	LONGS_EQUAL(0, 1/bar);
}

TEST(Separate_Process, thisTestCrashesWithCoreDump) {
	int* value = NULL;
	LONGS_EQUAL(15, *value);
}
```
